File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap17/TestREX.py
#!/usr/bin/env python

### TestREX.py

### Python version of Robert Cameron's REX XML shallow parser
### as posted by David Niergarth

import re                                                            #(A)

TextSE = "[^<]+"                                                     #(B)
UntilHyphen = "[^-]*-"                                               #(C)
Until2Hyphens = UntilHyphen + "(?:[^-]" + UntilHyphen + ")*-"        #(D)
CommentCE = Until2Hyphens + ">?"                                     #(E)
UntilRSBs = "[^\\]]*](?:[^\\]]+])*]+"                                #(F)
CDATA_CE = UntilRSBs + "(?:[^\\]>]" + UntilRSBs + ")*>"              #(G)
S = "[ \\n\\t\\r]+"                                                  #(H)
NameStrt = "[A-Za-z_:]|[^\\x00-\\x7F]"                               #(I)
NameChar = "[A-Za-z0-9_:.-]|[^\\x00-\\x7F]"                          #(J)
Name = "(?:" + NameStrt + ")(?:" + NameChar + ")*"                   #(K)
QuoteSE = "\"[^\"]*\"|'[^']*'"                                       #(L)
DT_IdentSE = S+Name+ "(?:" + S + "(?:" + Name + "|" + QuoteSE + "))*"#(M)
MarkupDeclCE = "(?:[^\\]\"'><]+|" + QuoteSE + ")*>"                  #(N)
S1 = "[\\n\\r\\t ]"                                                  #(O)
UntilQMs = "[^?]*\\?+"                                               #(P)
PI_Tail = "\\?>|" + S1 + UntilQMs + "(?:[^>?]" + UntilQMs + ")*>"    #(Q)
DT_ItemSE = "<(?:!(?:--" + Until2Hyphens + ">|[^-]" + MarkupDeclCE + \
        ")|\\?" + Name + "(?:" + PI_Tail + "))|%" + Name + ";|" + S  #(R)
DocTypeCE = DT_IdentSE + "(?:" + S + ")?(?:\\[(?:" + DT_ItemSE + \
                      ")*](?:" + S + ")?)?>?"                        #(S)
DeclCE = "--(?:" + CommentCE + ")?|\\[CDATA\\[(?:" + CDATA_CE + \
                      ")?|DOCTYPE(?:" + DocTypeCE + ")?"             #(T)
PI_CE = Name + "(?:" + PI_Tail + ")?"                                #(U)
EndTagCE = Name + "(?:" + S + ")?>?"                                 #(V)
AttValSE = "\"[^<\"]*\"|'[^<']*'"                                    #(W)
ElemTagCE = Name + "(?:" + S + Name + "(?:" + S + ")?=(?:" + S + \
                      ")?(?:" + AttValSE + "))*(?:" + S + ")?/?>?"   #(X)
MarkupSPE = "<(?:!(?:" + DeclCE + ")?|\\?(?:" + PI_CE + ")?|/(?:" + \
                      EndTagCE + ")?|(?:" + ElemTagCE + ")?)"        #(Y)
XML_SPE = TextSE + "|" + MarkupSPE                                   #(Z)

# Parsing function:
def ShallowParse(XML_document):                                      #(a)
	return re.findall(XML_SPE, XML_document)                     #(b)

if __name__ == '__main__':                                           #(c)
    import sys                                                       #(d)
    # Read the XML source document:
    FILE = open( sys.argv[1] )                                       #(e)
    doc_str = FILE.read()                                            #(f)
    parse_result = ShallowParse( doc_str )                           #(g)
    num_tokens = len(parse_result)                                   #(h)
    print "Number of tokens extracted: ",  num_tokens                #(i)
    for i in range(len(parse_result)):                               #(j)
        if parse_result[i]:                                          #(k)
            print "Item %s:  %s" % (i, parse_result[i])              #(l)
        else: print "item %s: " % i                                  #(m)

File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap17/TestREX.pl
#!/usr/bin/perl -w

### TestREX.pl

use strict;

# All of the regular expressions shown below that lead up to
# the definition of the XML shallow parsing expression that
# is assigned to $XML_SPE are by Robert Cameron [Cameron]

# Define the shallow parsing exression:

my $TextSE = "[^<]+";                                                #(A)
my $UntilHyphen = "[^-]*-";                                          #(B)
my $Until2Hyphens = "$UntilHyphen(?:[^-]$UntilHyphen)*-";            #(C)
my $CommentCE = "$Until2Hyphens>?";                                  #(D)
my $UntilRSBs = "[^\\]]*](?:[^\\]]+])*]+";                           #(E)
my $CDATA_CE = "$UntilRSBs(?:[^\\]>]$UntilRSBs)*>";                  #(F)
my $S = "[ \\n\\t\\r]+";                                             #(G)
my $NameStrt = "[A-Za-z_:]|[^\\x00-\\x7F]";                          #(H)
my $NameChar = "[A-Za-z0-9_:.-]|[^\\x00-\\x7F]";                     #(I)
my $Name = "(?:$NameStrt)(?:$NameChar)*";                            #(J)
my $QuoteSE = "\"[^\"]*\"|'[^']*'";                                  #(K)
my $DT_IdentSE = "$S$Name(?:$S(?:$Name|$QuoteSE))*";                 #(L)
my $MarkupDeclCE = "(?:[^\\]\"'><]+|$QuoteSE)*>";                    #(M)
my $S1 = "[\\n\\r\\t ]";                                             #(N)
my $UntilQMs = "[^?]*\\?+";                                          #(O)
my $PI_Tail = "\\?>|$S1$UntilQMs(?:[^>?]$UntilQMs)*>";               #(P)
my $DT_ItemSE = "<(?:!(?:--$Until2Hyphens>|[^-]$MarkupDeclCE)|" .  
                                "\\?$Name(?:$PI_Tail))|%$Name;|$S";  #(Q)
my $DocTypeCE = 
             "$DT_IdentSE(?:$S)?(?:\\[(?:$DT_ItemSE)*](?:$S)?)?>?";  #(R)
my $DeclCE = "--(?:$CommentCE)?|\\[CDATA\\[(?:$CDATA_CE)?|".
                                   "DOCTYPE(?:$DocTypeCE)?";         #(S)
my $PI_CE = "$Name(?:$PI_Tail)?";                                    #(T)
my $EndTagCE = "$Name(?:$S)?>?";                                     #(U)
my $AttValSE = "\"[^<\"]*\"|'[^<']*'";                               #(V)
my $ElemTagCE = "$Name(?:$S$Name(?:$S)?=(?:$S)?" .              
                                   "(?:$AttValSE))*(?:$S)?/?>?";     #(W)
my $MarkupSPE = "<(?:!(?:$DeclCE)?|\\?(?:$PI_CE)?|" .
                               "/(?:$EndTagCE)?|(?:$ElemTagCE)?)";   #(X)
my $XML_SPE = "$TextSE|$MarkupSPE";                                  #(Y)


# This function invoked in a list context will partition
# an XML document into its markup and text components:
sub ShallowParse {                                                   #(a)
  my($XML_document) = @_;                                            #(b)
  return $XML_document =~ /$XML_SPE/g;                               #(c)
}

# Read the XML source document:
$/ = undef;                                                          #(d)
my $string = <>;                                                     #(e)

my @parse_result = ShallowParse( $string );                          #(f)
my $num_tokens = @parse_result;                                      #(g)
print "Number of tokens extracted: $num_tokens\n\n";                 #(h)

my $i = 1;                                                     
foreach (@parse_result) {                                            #(i)
    print "Item $i:      $_\n" if  $_ !~ /^\s*$/;                    #(j)
    $i++;
}
